# Mini Project: Transfer Learning with Keras

Transfer learning is a machine learning technique where a model trained on one task is used as a starting point to solve a different but related task. Instead of training a model from scratch, transfer learning leverages the knowledge learned from the source task and applies it to the target task. This approach is especially useful when the target task has limited data or computational resources.

In transfer learning, the pre-trained model, also known as the "base model" or "source model," is typically trained on a large dataset and a more general problem (e.g., image classification on ImageNet, a vast dataset with millions of labeled images). The knowledge learned by the base model in the form of feature representations and weights captures common patterns and features in the data.

To perform transfer learning, the following steps are commonly followed:

1. Pre-training: The base model is trained on a source task using a large dataset, which can take a considerable amount of time and computational resources.

2. Feature Extraction: After pre-training, the base model is used as a feature extractor. The last few layers (classifier layers) of the model are discarded, and the remaining layers (feature extraction layers) are retained. These layers serve as feature extractors, producing meaningful representations of the data.

3. Fine-tuning: The feature extraction layers and sometimes some of the earlier layers are connected to a new set of layers, often called the "classifier layers" or "task-specific layers." These layers are randomly initialized, and the model is trained on the target task with a smaller dataset. The weights of the base model can be frozen during fine-tuning, or they can be allowed to be updated with a lower learning rate to fine-tune the model for the target task.

Transfer learning has several benefits:

1. Reduced training time and resource requirements: Since the base model has already learned generic features, transfer learning can save time and resources compared to training a model from scratch.

2. Improved generalization: Transfer learning helps the model generalize better to the target task, especially when the target dataset is small and dissimilar from the source dataset.

3. Better performance: By starting from a model that is already trained on a large dataset, transfer learning can lead to better performance on the target task, especially in scenarios with limited data.

4. Effective feature extraction: The feature extraction layers of the pre-trained model can serve as powerful feature extractors for different tasks, even when the task domains differ.

Transfer learning is commonly used in various domains, including computer vision, natural language processing (NLP), and speech recognition, where pre-trained models are fine-tuned for specific applications like object detection, sentiment analysis, or speech-to-text.

In this mini-project you will perform fine-tuning using Keras with a pre-trained VGG16 model on the CIFAR-10 dataset.

First, import all the libraries you'll need.

In [2]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, GlobalAveragePooling2D
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split
from tensorflow.keras.layers import Dropout
from tensorflow.keras.callbacks import EarlyStopping

The CIFAR-10 dataset is a widely used benchmark dataset in the field of computer vision and machine learning. It stands for the "Canadian Institute for Advanced Research 10" dataset. CIFAR-10 was created by researchers at the CIFAR institute and was originally introduced as part of the Neural Information Processing Systems (NIPS) 2009 competition.

The dataset consists of 60,000 color images, each of size 32x32 pixels, belonging to ten different classes. Each class contains 6,000 images. The ten classes in CIFAR-10 are:

1. Airplane
2. Automobile
3. Bird
4. Cat
5. Deer
6. Dog
7. Frog
8. Horse
9. Ship
10. Truck

The images are evenly distributed across the classes, making CIFAR-10 a balanced dataset. The dataset is divided into two sets: a training set and a test set. The training set contains 50,000 images, while the test set contains the remaining 10,000 images.

CIFAR-10 is often used for tasks such as image classification, object recognition, and transfer learning experiments. The relatively small size of the images and the variety of classes make it a challenging dataset for training machine learning models, especially deep neural networks. It also serves as a good dataset for teaching and learning purposes due to its manageable size and straightforward class labels.

Here are your tasks:

1. Load the CIFAR-10 dataset after referencing the documentation [here](https://keras.io/api/datasets/cifar10/).
2. Normalize the pixel values so they're all in the range [0, 1].
3. Apply One Hot Encoding to the train and test labels using the [to_categorical](https://www.tensorflow.org/api_docs/python/tf/keras/utils/to_categorical) function.
4. Further split the the training data into training and validation sets using [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html). Use only 10% of the data for validation.  

In [5]:
# Load the CIFAR-10 dataset

(x_train, y_train), (x_test, y_test) = cifar10.load_data()

In [6]:
# Normalize the pixel values to [0, 1]
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

In [7]:
# One-hot encode the labels
y_train = to_categorical(y_train, num_classes=10)
y_test = to_categorical(y_test, num_classes=10)

In [8]:
# Split the data into training and validation sets
x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.1, random_state=42)

VGG16 (Visual Geometry Group 16) is a deep convolutional neural network architecture that was developed by the Visual Geometry Group at the University of Oxford. It was proposed by researchers Karen Simonyan and Andrew Zisserman in their paper titled "Very Deep Convolutional Networks for Large-Scale Image Recognition," which was presented at the International Conference on Learning Representations (ICLR) in 2015.

The VGG16 architecture gained significant popularity for its simplicity and effectiveness in image classification tasks. It was one of the pioneering models that demonstrated the power of deeper neural networks for visual recognition tasks.

Key characteristics of the VGG16 architecture:

1. Architecture: VGG16 consists of a total of 16 layers, hence the name "16." These layers are stacked one after another, forming a deep neural network.

2. Convolutional Layers: The main building blocks of VGG16 are the convolutional layers. It primarily uses 3x3 convolutional filters throughout the network, which allows it to capture local features effectively.

3. Max Pooling: After each set of convolutional layers, VGG16 applies max-pooling layers with 2x2 filters and stride 2, which halves the spatial dimensions (width and height) of the feature maps and reduces the number of parameters.

4. Fully Connected Layers: Towards the end of the network, VGG16 has fully connected layers that act as a classifier to make predictions based on the learned features.

5. Activation Function: The network uses the Rectified Linear Unit (ReLU) activation function for all hidden layers, which helps with faster convergence during training.

6. Number of Filters: The number of filters in each convolutional layer is relatively small compared to more recent architectures like ResNet or InceptionNet. However, stacking multiple layers allows VGG16 to learn complex hierarchical features.

7. Output Layer: The output layer consists of 1000 units, corresponding to 1000 ImageNet classes. VGG16 was originally trained on the large-scale ImageNet dataset, which contains millions of images from 1000 different classes.

VGG16 was instrumental in showing that increasing the depth of a neural network can significantly improve its performance on image recognition tasks. However, the main drawback of VGG16 is its high number of parameters, making it computationally expensive and memory-intensive to train. Despite this limitation, VGG16 remains an essential benchmark architecture and has paved the way for even deeper and more efficient models in the field of computer vision, such as ResNet, DenseNet, and EfficientNet.

Here are your tasks:

1. Load [VGG16](https://keras.io/api/applications/vgg/#vgg16-function) as a base model. Make sure to exclude the top layer.
2. Freeze all the layers in the base model. We'll be using these weights as a feature extraction layer to forward to layers that are trainable.

In [9]:
# Load the pre-trained VGG16 model (excluding the top classifier)
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))

58889256/58889256 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [10]:
# Freeze the layers in the base model
for layer in base_model.layers:
    layer.trainable = False

Now, we'll add some trainable layers to the base model.

1. Using the base model, add a [GlobalAveragePooling2D](https://keras.io/api/layers/pooling_layers/global_average_pooling2d/) layer, followed by a [Dense](https://keras.io/api/layers/core_layers/dense/) layer of length 256 with ReLU activation. Finally, add a classification layer with 10 units, corresponding to the 10 CIFAR-10 classes, with softmax activation.
2. Create a Keras [Model](https://keras.io/api/models/model/) that takes in approproate inputs and outputs.

In [11]:
# Add a global average pooling layer
x = base_model.output
x = GlobalAveragePooling2D()(x)

In [12]:
# Add a fully connected layer with 256 units and ReLU activation
x = Dense(256, activation='relu')(x)

In [13]:
# Add the final classification layer with 10 units (for CIFAR-10 classes) and softmax activation
predictions = Dense(10, activation='softmax')(x)

In [14]:
# Create the fine-tuned model
model = Model(inputs=base_model.input, outputs=predictions)

With your model complete it's time to train it and assess its performance.

1. Compile your model using an appropriate loss function. Feel free to play around with the optimizer, but a good starting optimizer might be Adam with a learning rate of 0.001.
2. Fit your model on the training data. Use the validation data to print the accuracy for each epoch. Try training for 10 epochs. Note, training can take a few hours so go ahead and grab a cup of coffee.

**Optional**: See if you can implement an [Early Stopping](https://keras.io/api/callbacks/early_stopping/) criteria as a callback function.

In [12]:
# Compile the model
model.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
# Train the model
history = model.fit(x_train, y_train, batch_size=64, epochs=10, validation_data=(x_val, y_val))

Epoch 1/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 531s 753ms/step - accuracy: 0.4506 - loss: 1.5766 - val_accuracy: 0.5608 - val_loss: 1.2429
Epoch 2/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 530s 708ms/step - accuracy: 0.5766 - loss: 1.2125 - val_accuracy: 0.5792 - val_loss: 1.1943
Epoch 3/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 534s 754ms/step - accuracy: 0.6005 - loss: 1.1404 - val_accuracy: 0.5938 - val_loss: 1.1538
Epoch 4/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 559s 749ms/step - accuracy: 0.6142 - loss: 1.0982 - val_accuracy: 0.5990 - val_loss: 1.1363
Epoch 5/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 561s 749ms/step - accuracy: 0.6323 - loss: 1.0537 - val_accuracy: 0.6116 - val_loss: 1.1138
Epoch 6/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 530s 704ms/step - accuracy: 0.6464 - loss: 1.0123 - val_accuracy: 0.6180 - val_loss: 1.1051
Epoch 7/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 533s 748ms/step - accuracy: 0.6605 - loss: 0.9746 - val_accuracy: 0.6212 - val_loss: 1.0870
Epoch 8/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 495s 703ms/step - accuracy: 0.6671 -

With your model trained, it's time to assess how well it performs on the test data.

1. Use your trained model to calculate the accuracy on the test set. Is the model performance better than random?
2. Experiment! See if you can tweak your model to improve performance.  

In [14]:
# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Loss: {test_loss:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")
print(f"Accuracy percentage: {test_accuracy * 100: .2f}%")


313/313 ━━━━━━━━━━━━━━━━━━━━ 101s 322ms/step - accuracy: 0.6064 - loss: 1.1379
Test Loss: 1.1413
Test Accuracy: 0.6065
Accuracy percentage:  60.65%


In [15]:
# Test at random
y_test = y_test.flatten()

random = np.random.randint(0, 10, size=y_test.shape)

accuracy = np.mean(random == y_test)
print(f"Simulated random guessing accuracy: {accuracy * 100:.2f}%")

Simulated random guessing accuracy: 10.01%


**Question:** Use your trained model to calculate the accuracy on the test set. Is the model performance better than random?

**Answer:** The model performance at __ is better than random, 10.04%.

In [15]:
# Attempt to increase performance and accuracy: Unfreeze the model to help fine tune
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(256, activation='relu')(x)
predictions = Dense(10, activation='softmax')(x)

model = Model(inputs=base_model.input, outputs=predictions)
base_model.trainable = True

model.compile(optimizer=Adam(learning_rate=1e-5),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit( x_train, y_train, batch_size=64, epochs=10, validation_data=(x_val, y_val))

# Evaluate on the test set
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Loss: {test_loss:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")
print(f"Accuracy percentage: {test_accuracy * 100:.2f}%")


Epoch 1/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 560s 397ms/step - accuracy: 0.1623 - loss: 2.3103 - val_accuracy: 0.3278 - val_loss: 1.9970
Epoch 2/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 559s 395ms/step - accuracy: 0.3618 - loss: 1.9380 - val_accuracy: 0.4082 - val_loss: 1.8062
Epoch 3/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 557s 396ms/step - accuracy: 0.4220 - loss: 1.7720 - val_accuracy: 0.4464 - val_loss: 1.6878
Epoch 4/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 529s 372ms/step - accuracy: 0.4524 - loss: 1.6679 - val_accuracy: 0.4604 - val_loss: 1.6103
Epoch 5/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 562s 372ms/step - accuracy: 0.4757 - loss: 1.5937 - val_accuracy: 0.4742 - val_loss: 1.5558
Epoch 6/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 594s 395ms/step - accuracy: 0.4849 - loss: 1.5460 - val_accuracy: 0.4846 - val_loss: 1.5160
Epoch 7/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 562s 395ms/step - accuracy: 0.4951 - loss: 1.5089 - val_accuracy: 0.4938 - val_loss: 1.4842
Epoch 8/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 562s 395ms/step - ac

In [16]:
# Attempt to increase performance and accuracy: add dropout to try to reduce overfitting
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))
base_model.trainable = False

x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(256, activation='relu')(x)
x = Dropout(0.5)(x)
predictions = Dense(10, activation='softmax')(x)
model = Model(inputs=base_model.input, outputs=predictions)
model.compile(optimizer=Adam(learning_rate=0.001),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit( x_train, y_train, batch_size=64, epochs=10, validation_data=(x_val, y_val))

# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Loss: {test_loss:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")
print(f"Accuracy percentage: {test_accuracy * 100:.2f}%")


Epoch 1/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 515s 732ms/step - accuracy: 0.5257 - loss: 1.4115 - val_accuracy: 0.5176 - val_loss: 1.4086
Epoch 2/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 542s 770ms/step - accuracy: 0.5229 - loss: 1.4043 - val_accuracy: 0.5202 - val_loss: 1.4001
Epoch 3/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 562s 770ms/step - accuracy: 0.5338 - loss: 1.3851 - val_accuracy: 0.5204 - val_loss: 1.3910
Epoch 4/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 537s 735ms/step - accuracy: 0.5340 - loss: 1.3824 - val_accuracy: 0.5230 - val_loss: 1.3814
Epoch 5/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 557s 728ms/step - accuracy: 0.5426 - loss: 1.3670 - val_accuracy: 0.5238 - val_loss: 1.3731
Epoch 6/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 544s 773ms/step - accuracy: 0.5381 - loss: 1.3602 - val_accuracy: 0.5280 - val_loss: 1.3652
Epoch 7/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 528s 726ms/step - accuracy: 0.5397 - loss: 1.3545 - val_accuracy: 0.5296 - val_loss: 1.3580
Epoch 8/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 562s 726ms/step - accuracy: 0.5445 -

In [17]:
# Attempt to increase performance and accuracy: add Early Stopping
x = base_model.output
x = GlobalAveragePooling2D()(base_model.output)
x = Dense(256, activation='relu')(x)
predictions = Dense(10, activation='softmax')(x)

early_stop = EarlyStopping(
    monitor='val_loss',
    min_delta=0.001,
    patience=5,
    verbose=1,
    mode="auto",
    restore_best_weights=True
)

model.fit(x_train, y_train, batch_size=64, epochs=10, validation_data=(x_val, y_val), callbacks=[early_stop])

# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Loss: {test_loss:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")
print(f"Accuracy percentage: {test_accuracy * 100:.2f}%")

Epoch 1/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 530s 752ms/step - accuracy: 0.5494 - loss: 1.3278 - val_accuracy: 0.5412 - val_loss: 1.3334
Epoch 2/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 579s 776ms/step - accuracy: 0.5522 - loss: 1.3222 - val_accuracy: 0.5424 - val_loss: 1.3272
Epoch 3/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 556s 767ms/step - accuracy: 0.5572 - loss: 1.3082 - val_accuracy: 0.5458 - val_loss: 1.3225
Epoch 4/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 536s 761ms/step - accuracy: 0.5550 - loss: 1.3078 - val_accuracy: 0.5440 - val_loss: 1.3175
Epoch 5/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 563s 763ms/step - accuracy: 0.5609 - loss: 1.2954 - val_accuracy: 0.5468 - val_loss: 1.3125
Epoch 6/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 505s 717ms/step - accuracy: 0.5602 - loss: 1.2968 - val_accuracy: 0.5476 - val_loss: 1.3080
Epoch 7/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 563s 718ms/step - accuracy: 0.5620 - loss: 1.2905 - val_accuracy: 0.5508 - val_loss: 1.3040
Epoch 8/10
704/704 ━━━━━━━━━━━━━━━━━━━━ 592s 761ms/step - accuracy: 0.5636 -

**Attempt to increase performance and accuracy:**

From above 3 attempts were made to increase performance and accuracy. These attempts were: (1) to unfreeze the model, (2) to add dropout from keras, and (3) to add early stopping as recommended.


 (1) Allows their weights or how the model is calculating the patterns to adjust accordingly instead of keeping them fixed.

 (2) Drops neurons so that the model doesn't rely on any one neural path which can lead to overfitting. Therefore, this will make better generalization.

 (3) Stops the model before overfitting can happen.

 Out of the 3 we can see that early stopping had the best performance with an accuracy at 54.83%. This is followed by dropout which has an accuracy at 53.67%. And lastly, unfreezing the model which has an accuracy at 51.29%.

 However, all of these techniques did not improve the original model which had accuracy of 60.65%.